src/workerd/api/container.c++

```diff
@@ -48,18 +48,74 @@ void Container::start(jsg::Lock& js, jsg::Optional<StartupOptions> maybeOptions)
   }
 
   IoContext::current().addTask(req.sendIgnoringResult());
-
   running = true;
+  monitorOnBackgroundIfNeeded();
 }
 
-jsg::Promise<void> Container::monitor(jsg::Lock& js) {
+void Container::monitorOnBackgroundIfNeeded() {
+  if (monitoring || !running) {
+    return;
+  }
+  monitoring = true;
+
+  if (monitorKjPromise == kj::none) {
+    // Store the exit code before forking since Response is not copyable
+    monitorKjPromise = rpcClient->monitorRequest(capnp::MessageSize{4, 0})
+                           .send()
+                           .then([](auto&& results) -> uint8_t {
+      return results.getExitCode();
+    }).fork();
+  }
+
+  auto req = KJ_ASSERT_NONNULL(monitorKjPromise)
+                 .addBranch()
+                 .then([this, self = JSG_THIS](uint8_t exitCode) {
+    running = false;
+    monitoring = false;
```

Comment:
You can likely get away with using an `.attach(kj::defer([this, self = JSG_THIS]() { running = false; monitoring = false; }))` to the promise rather than explicitly handling it in both the then and error handler to eliminate the duplication here.